feat(lesson): pep8, linters & code formatters (#12)
pyOpenSci/lessons: clean-modular-code/python-pep-8.md

```{code-cell} ipython3
---
editable: true
slideshow:
  slide_type: ''
---
#some data analysis with poor formatting
import pandas as pd
from datetime import datetime
def classify_precipitation(precip_list):
 avg_precip=pd.Series(precip_list).mean()
 if avg_precip<100:
  return'Low'
 elif avg_precip>=100 and avg_precip<=150:
  return'Medium'
 else:
  return'High'
data={'location':['Station1','Station2','Station3','Station4'],'year':[2021,2021,2021,2021],'monthly_precipitation':[[50.0,70.0,90.0,80.0],[100.0,110.0,120.0,130.0],[150.0,160.0,170.0,180.0],[200.0,210.0,220.0,230.0]],'start_date':["2021-01-01","2021-01-01","2021-01-01","2021-01-01"]}
df=pd.DataFrame(data)
df['start_date']=pd.to_datetime(df['start_date'])
df['precipitation_category']=df['monthly_precipitation'].apply(classify_precipitation)
df
```

+++ {"editable": true, "slideshow": {"slide_type": ""}}

Notice that by adding docstrings, spacing to the the code.... 

> ```{code-cell} ipython3

Reviewer: `E501` is something that I always suppress as unneeded in modern code.

> ```{code-cell} ipython3

Reviewer 2: Thats reassuring to know! I usually bump it up to 90 :)

> ```{code-cell} ipython3

Author: Oh wow—I always use that 79-character limit. It's just a habit at this point—all of my text editors are set up that way. But what is annoying is when it's applied to docstrings. I never thought about it as "optional," but I guess it is!

    if avg_precip < 100:
        return 'Low'
    elif 100 <= avg_precip <= 150:
        return 'Medium'
    else:
        return 'High'

# Sample data for precipitation values (in mm) for different locations
data = {
    'location': ['Station1', 'Station2', 'Station3', 'Station4'],
    'year': [2021, 2021, 2021, 2021],
    'monthly_precipitation': [
        [50.0, 70.0, 90.0, 80.0],  # Station1
        [100.0, 110.0, 120.0, 130.0],   # Station2
        [150.0, 160.0, 170.0, 180.0],   # Station3
        [200.0, 210.0, 220.0, 230.0]    # Station4
    ],
    'start_date': ["2021-01-01", "2021-01-01", "2021-01-01", "2021-01-01"]
}

df = pd.DataFrame(data)
df['start_date'] = pd.to_datetime(df['start_date'])

# Classify precipitation levels based on ranges
df['precipitation_category'] = df['monthly_precipitation'].apply(classify_precipitation)

df
```

+++ {"editable": true, "slideshow": {"slide_type": ""}}

## How to Apply PEP 8 Code Style Standards

It may seem overwhelming to remember all the PEP 8 rules, but tools called **code formatters** can automatically apply these standards for you.

### For `.py` Files

Use popular tools like **Black** or **Ruff**:

> - **[Black](https://black.readthedocs.io/en/stable/)**: Automatically reformats code according to PEP 8.

Reviewer: As a co-maintainer of Black, the original intent was to apply rules to minimize the size of diffs and to make code review easier. 

@sneakers-the-rat I would be interested when black and ruff contradict each other.

> - **[Black](https://black.readthedocs.io/en/stable/)**: Automatically reformats code according to PEP 8.

Reviewer 2: That's the main one I have experienced! Mostly when writing `@override`s, black will want to put the ellipsis on the same line and ruff will want to put it on the following line, its not a big deal I just have to remember to run black after ruff which is what I think you're supposed to do anyway?

> - **[Black](https://black.readthedocs.io/en/stable/)**: Automatically reformats code according to PEP 8.

Author: So, I've stopped using Ruff on some projects and have migrated to Ruff for code formatting, too, for ease of tooling. But i still use black as a code formatter in vscode. It's so interesting that there are so many different options when thinking about linting and code formatting tools. I say this because only a few years ago i was using autopep8 for software and the autopep8 tool for jupyter in teaching.

Using these tools ensures your code remains consistent, readable, and compliant with PEP 8, without memorizing all the rules.

:::{admonition} Linter vs. Code Formatter
:class: note

- **Linter**: Checks your code and highlights issues but doesn’t automatically fix them.
- **Code Formatter**: Automatically reformats your code according to style rules.

Ruff acts as both a linter and formatter, making it ideal for `.py` file workflows.
:::

+++ {"editable": true, "slideshow": {"slide_type": ""}}

## Why use code standards when writing Python code

Code standards make your code more readable and easier to maintain, just like writing conventions make text easier to read. Think about how we capitalize the first word of a sentence or add spaces between words—without those conventions, text would be hard to follow. Similarly, code with inconsistent formatting is difficult to understand and debug.
  
For example:

* Readable sentence: This is a sentence.
* Unformatted sentence: thisisasentence.withoutspacesitgetsconfusing.

Using code standards like PEP 8 helps avoid such confusion, making code clearer and more professional.

### Some PEP 8 rules to remember 

> There is a style guide devoted to Python pep 8 standards that you can read [here](https://www.python.org/dev/peps/pep-0008/#naming-conventions). However, below are a handful of PEP 8 Rules that you can consider following when writing code. 

Reviewer: ```suggestion
There is a style guide devoted to Python PEP 8 standards that you can read [here](https://www.python.org/dev/peps/pep-0008/#naming-conventions). However, below are a handful of PEP 8 Rules that you can consider following when writing code. 
```

* Naming Conventions: Use **snake_case** for variables/functions and **CamelCase** for class names.

```python
# This is a class definition
class MyClass:
    """A class to process data and calculate statistics."""

    # This is a function 
    def calculate_average(self, data_list):
        """Calculate the average of a list of numbers."""
        total_sum = sum(data_list)
        count = len(data_list)
        return total_sum / count

> # Example variable names and function call
> data_processor = DataProcessor()
> numbers = [10, 20, 30, 40, 50]
> average_value = data_processor.calculate_average(numbers)

Reviewer: ```suggestion
# Call to create an object of a class
data_processor = DataProcessor()
numbers = [10, 20, 30, 40, 50]
# Examples of variable names and method call
average_value = data_processor.calculate_average(numbers)
```

> 

Reviewer: ```suggestion
# Example of function call
```

print(f"The average value is: {average_value}")
```

* Line Length: Limit all lines to 79 characters for better readability.

:::{tip}
Most text editors allow you to set up guides to see how long your code is. You can then use these guides to create line breaks in your code.
:::

* Comments: Add a space after the `#` sign and capitalize the first letter of a comment:

`# This is a PEP 8 conforming comment`

> * White Space: Add space between sections of code to improve clarity.

Reviewer: ```suggestion
* White Space: Add space between sections of code to improve readability.
```

* Avoid using single-character letters that could be confused with numbers  (ie the letter `l` looks similar to the number one  `1` 

> * Add a blank line before a single-line comment (unless it is the first line of a cell in Jupyter Notebook)

Reviewer: ```suggestion
* Add a blank line before a single-line comment (unless it is the first line of a cell in Jupyter Notebook or part of a code block)
```

```python
a = 1 

# Here is a commment
b = 2 

```

> * **Break up sections of code with white space:** Breaking up your

Reviewer 2: an idea: all the whitespace stuff could go in a single example like the capitalization stuff

```python
from itertools import counter

# two lines above a function definition
def my_function():
    count = counter()
    values = [next(count) for i in range(10)]

    # one line to break up code into related chunks
    plt.plot(values)

# Two lines before a class definition
class MyClass:
    
    # one line before a method definition
    def my_method(): ...
```

code becomes even more important when you start working in Jupyter Notebooks which offer individual cells where you can add Markdown and code.

```python
# Process some data here 

> data=pd.readcsv("pyos-data.csv")

Reviewer 2: it won't let me comment on line 231 but i think that should be `plt.subplots()`

# Plot data - notice how separating code into sections makes it easier to read
fig, ax = plot.subplots()
data.plot(ax=ax)
plt.show()
```

> ## PEP 8 naming conventions

Reviewer 2: this is odd to me to have as its own top-level section after listing a bunch of pep 8 rules in the previous subsection, and it seems to be the same information as the example starting on line 179? i would combine those things and put them under a section on various types of conventions. i'll give an example ToC in the main comment

:::{seealso}

> For the entire pep-8 style guide see: <a href="https://peps.python.org/pep-0008/" target="_blank">PEP 8 Style Guide published by the Python Software Foundation</a>.

Reviewer: This is more accurate.

```suggestion
For the entire pep-8 style guide see: <a href="https://peps.python.org/pep-0008/" target="_blank">Python's PEP 8 Style Guide</a>.
```

:::

:::{admonition} Terminology Review
:class: tip

> First, let's review some terminology associated with naming conventions.

Reviewer 2: section starts "first" but there is no next!

* **mixedCase:** (differs from CapitalizedWords by initial lowercase character!)
* **Capitalized_Words_With_Underscores:** This approach is not recommended. Use one convention and stick with it.

:::

+++ {"editable": true, "slideshow": {"slide_type": ""}}

## Best practices for importing libraries

### Import Python libraries at the top of your code

It’s good practice to import all required libraries at the top of your **Python** script or in the first cell of a **Jupyter Notebook**. This helps anyone understand what packages are needed to run your code. It also follows PEP 8 conventions.

### Organize your imports into groups

PEP 8 recommends organizing imports in the following order:

1. **Standard Library Imports**: These built-in modules come with Python, such as `os` and `glob`. You can find the full list [here](https://docs.python.org/3/library/index.html).
2. **Third-Party Imports**: Libraries that you install via `pip`, like `numpy` and `pandas`.
3. **Local Imports**: Code or modules specific to your project.

Here’s an example following PEP 8 conventions:

```python

> import os

Reviewer: Let's alphabetize these imports.

Reviewer 2: isort makes it a little hard to know what the rules are exactly but if we want to add the sorting rules in here, here are the most common rules (in addition to grouping and alphabetization):

- modules:
  - module-only imports (`import x`) come before imports with objects (`from x import y`)
  - within that, import statements sorted alphabetically by module name
- objects within a module import (the `y` in `from x import y`)
  - first sort by whether the first letter is capitalized or not
  - then alphabetically

import glob

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from my_module import my_function
```

### Why organize imports?

> Organizing your imports this way ensures your code is readable and follows widely accepted practices. Importing libraries at the top also makes it easier to debug and see which dependencies are required to run the code.

Reviewer: ```suggestion
Organizing your imports this way ensures your code is readable and follows widely accepted practices. Importing libraries at the top also makes it easier to scan for a specific library, debug dependencies, and see which dependencies are required to run the code.
```

> Organizing your imports this way ensures your code is readable and follows widely accepted practices. Importing libraries at the top also makes it easier to debug and see which dependencies are required to run the code.

Reviewer 2: and merge conflicts, but i think basically we should consolidate "why code style" at the beginning because it's mostly the same reasons

```python
import os 

import pandas as pd
import numpy as np
```

Once you have imported all of the packages that you need to run your code in a script, you have access to all of the functions and classes defined in each package. 

If these imports are at the top of the script or **Jupyter Notebook** file, then you will be able to use those packages in any code lines that follow. 

This means that if you import a package *after* running some code that requires that package, your code will not run successfully.

## 2. List Package Imports Following PEP 8 Standards: Most Common First, Followed By Third Party

<a href="https://www.python.org/dev/peps/pep-0008/#imports" target="_blank">PEP 8</a> also specifies the order in which you should list your imports as follows (starting with the most commonly used):

> Imports should be grouped in the following order:
>    Standard library imports.
>    Related third party imports.
>    Local application/library specific imports.
> You should put a blank line between each group of imports.

> You may be wondering, what is a standard library import? The standard library imports are commonly used tools that are general in purpose and are part of the standard library of **Python**. These including things like:

Reviewer 2: "that are included with python," "that you don't have to install," with link to https://docs.python.org/3/library/index.html ?

* **os**: handle files and directories.
* **glob**: create lists of files and directories for batch processing.

In the PEP 8 order, other commonly used packages that are general in purpose will follow such as: 

* **numpy**: work with data in array formats.
* **matplotlib**: plot data.
* * **Pandas**: to work with tabular data.

A PEP 8 order of imports for commonly used **Python** packages for science would look something like this:

```python
import os
import glob

import matplotlib.pyplot as plt
import numpy as np

> import pandas as pd

Reviewer 2: add the third block of 

```python
# imports from your local package
import mypackage
from mypackage.internal import _secret_function
```

Reviewer 2: oh i see you have a separate section for that. my b. i just figured showing all three of them in the example that follows the list of 3 would be nice, but this is a draft and i'm jumping the gun on giving comments lol

Author: go for it!! comment away. i just combined two pages. i'm totally fine with consolidating!!  

 This is good, too, because in our second workshop, they will create a package and import it. so they will get to try this out!!

Author: I've just pushed a new version of this and made all of the fixes that you suggested. I also redesigned the structure to be:

* Example of good vs not as good code format. 
* How to apply tools to help you make your code prettier :) 
* Why use these tools and standards
* Some rules to remember 

That way, we start with something that doesn't have to be so hard and end with the details that many people will skip (I think?).

Reviewer 2: I hear you on that but i think that currently it's a little bewildering the other way where we're missing some introductory information on what we're talking about and why first, about to post the review

edit: oh wait it makes this comment part of the review even when we started it before that lol my b

Author: 😆 no worries!! i think / hope that i addressed your comment!! I am going to focus on merging this today give we are a week out BUT we can always revisit all of this content and update/ enhance / beautify, etc in the future!! 🚀 

thank you for this feedback!!
